src/Package/Impl/DataInspect/DataImport/ImportDataWindow.xaml.cs

```diff
@@ -0,0 +1,304 @@
+            builder.AppendFormat("\"{0}\"", value);
+        }
+
+        private void FileOpenButton_Click(object sender, RoutedEventArgs e) {
+            IVsUIShell uiShell = VsAppShell.Current.GetGlobalService<IVsUIShell>(typeof(SVsUIShell));
+            IntPtr dialogOwner;
+            uiShell.GetDialogOwnerHwnd(out dialogOwner);
+
+            FilePathBox.Text = Dialogs.BrowseForFileOpen(dialogOwner, "CSV file|*.csv");
+            if (!string.IsNullOrEmpty(FilePathBox.Text)) {
+                FilePathBox.CaretIndex = FilePathBox.Text.Length;
+                FilePathBox.ScrollToEnd();
+
+                var variableName = System.IO.Path.GetFileNameWithoutExtension(FilePathBox.Text);
+                VariableNameBox.Text = variableName;
+
+                string text = ReadFile(FilePathBox.Text);
+                InputFilePreview.Text = text;
+
+                PreviewAsync().DoNotWait();
+            }
+        }
+
+        private void RunButton_Click(object sender, RoutedEventArgs e) {
+            RunButton.IsEnabled = false;
+            CancelButton.IsEnabled = false;
+
+            var expression = BuildCommandLine(false);
+            if (expression != null) {
+                RunAsync(expression).DoNotWait();
+            }
+        }
+
+        private void CancelButton_Click(object sender, RoutedEventArgs e) {
+            base.Close();
+        }
+
+        private async Task PreviewAsync() {
+            var expression = BuildCommandLine(true);
+            if (expression != null) {
+                try {
+                    var grid = await GridDataSource.GetGridDataAsync(expression, null);
+
+                    VsAppShell.Current.DispatchOnUIThread(() => {
+                        PopulateDataFramePreview(grid);
+                        OnSuccess(true);
+                    });
+                } catch (Exception ex) {
+                    VsAppShell.Current.DispatchOnUIThread(() => OnError(ex.Message));
+                }
+            }
+        }
+
+        private async Task RunAsync(string expression) {
+            await TaskUtilities.SwitchToBackgroundThread();
+
+            try {
+                IRSession rSession = GetRSession();
+
+                using (var evaluator = await rSession.BeginEvaluationAsync()) {
+                    var result = await evaluator.EvaluateAsync(expression);
+                    if (result.ParseStatus != RParseStatus.OK || result.Error != null) {
+                        VsAppShell.Current.DispatchOnUIThread(() => OnError(result.ToString()));
+                    } else {
+                        VsAppShell.Current.DispatchOnUIThread(() => OnSuccess(false));
+                    }
+                }
+            } catch (Exception ex) {
+                VsAppShell.Current.DispatchOnUIThread(() => OnError(ex.Message));
+            }
+        }
+
+        private async Task SetEncodingComboBoxAsync() {
+            await TaskUtilities.SwitchToBackgroundThread();
```

Comment:
This method will be easier to read if you move to background thread only `BeginEvaluationAsync` part. In that case, you wouldn't have to call `DispatchOnUIThread`:
```{csharp}
private async Task SetEncodingComboBoxAsync() {
    try {
        var result = await CallIconvlist();
        if (result.ParseStatus != RParseStatus.OK || result.Error != null) {
            OnError(result.ToString());
        } else {
            var jarray = result.JsonResult as JArray;
            if (jarray != null) {
                PopulateEncoding(jarray);
            }
        }
    } catch (Exception ex) {
        OnError(ex.Message);
    }
}

private async Task<REvaluationResult> CallIconvlist() {
    await TaskUtilities.SwitchToBackgroundThread();
    var rSession = GetRSession();
    using (var evaluator = await rSession.BeginEvaluationAsync(false)) {
        return await evaluator.EvaluateAsync("rtvs:::toJSON(as.list(iconvlist()))", REvaluationKind.Json);
    }
}
```